### PR TITLE
RFC: Add `uki` layout to kernel-install

### PR DESCRIPTION
**Pull request by behrmann**
Currently the kernel-install man page only documents the bls layout for use with the boot loader spec type #1. `90-loaderentry.install` uses this layout to generate loader entries and copy the kernel image and initrd to `$BOOT`.

This PR documents a second layout `uki` and adds `91-ukiinstall.install`, which copies a UKI given on the command line or from the staging area to `$BOOT/EFI/Linux/$ENTRY_TOKEN-$KERNEl_VERSION(+$TRIES).efi`

### CODE REVIEW

**Comment by poettering** [comment]
hmm, so if "kernel-install xyz abc" is called, and KERNEL_INSTALL_LAYOUT=uki is set then abc would have to refer to a proper UKI, but if the same command is set and KERNEL_INSTALL_LAYOUT=bls is set then to a regular kernel?

i must say i find this a bit icky that the cmdline params need to be different based on configuration, in particular as the caller of `kernel-install` probably shouldn't be agnostic to the install layout...

My assumption was always that once we add something like this, we'd ship a plugin script that would be capable of locally building ukis from the passed params, i.e. along the lines of #25180 

@keszybz what's your take on this?

I'd feel more comfortable, if we'd have some check in place that generates a clean error and fails if the kernel image passed is not a proper UKI. So that it doesn't happen by accident that for example on Fedora regular kernel rpms installing non-UKI kernels end up installing their images as UKIs... Not sure how to do that best though.

**Comment by behrmann** [comment]
Ah, sorry, I forgot to put that that part of the reasoning into the commit message. 

I do think, that we should support UKIs as file input for `kernel-install` at some point, but this here is meant slightly differently. I'd actually like to make this usuable for two cases:
1) distros shipping UKIs
2) people/organisations on distros not (yet) shipping UKIs and wanting to generate them locally. 

We already have the concept of `KERNEL_INSTALL_GENERATOR` to choose which initrd generator runs locally (if any). The idea would be with a layout `uki` the generator can act on that an generate a UKI and put it into the staging area from whence it is then copied to `$BOOT` by `91-ukiinstall.install`. 

Distros shipping UKIs and wishing to use `kernel-install` could just put some `50-stage-uki.install` file into `/usr/lib/kernel/install.d` to get the UKI in the staging area.

I guess it would be clearer to just drop everything but `KERNEL_INSTALL_STAGING_AREA` from where to read UKIs - at least for now.

The question arose from [a discussion on the mkinitcpio bugtracker](https://gitlab.archlinux.org/archlinux/mkinitcpio/mkinitcpio/-/issues/138), since `kernel-installs`'s default `bls` layout targets BLS type 1 and I think it would be nice to have a common layout that different generators interested in UKIs could use. 

/cc @Foxboron

**Comment by keszybz** [comment]
> hmm, so if "kernel-install xyz abc" is called, and KERNEL_INSTALL_LAYOUT=uki is set then abc would have to refer to a proper UKI, but if the same command is set and KERNEL_INSTALL_LAYOUT=bls is set then to a regular kernel?
> 
> i must say i find this a bit icky that the cmdline params need to be different based on configuration, in particular as the caller of `kernel-install` probably shouldn't be agnostic to the install layout...

The caller of `kernel-install` generally isn't agnostic of the install layout. The caller can be either a developer (more likely for now), or a package scriptlet. The package would generally "know" what type is being installed, because you'd have separate packages for uki and non-uki kernels (if both are packaged), because of the size. So the scriptlet can pass a different argument or add some additional options as desired. If we're installing an UKI in this way, we would want to pass the path to the UKI as a argument, to avoid any confusion if there's more than one, so using that as a way to distinguish what is wanted seems natural.

 > My assumption was always that once we add something like this, we'd ship a plugin script that would be capable of locally building ukis from the passed params, i.e. along the lines of #25180

We'll want both. At some point distros might start shipping pre-packaged UKIs (e.g. https://fedoraproject.org/wiki/Changes/Unified_Kernel_Support_Phase_1), but we'll also want to assemble them locally. When we discussed this with @behrmann, the general idea was that a *different* plugin would generate the UKI into the staging area, and *this* plugin would then install the file into the destination. That's why it already looks in the staging area.

> I guess it would be clearer to just drop everything but `KERNEL_INSTALL_STAGING_AREA` from where to read UKIs - at least for now.

No, please don't do this. Packaged UKIs are already appearing (see the linked copr repo in the abovementioned proposal), and I don't want people to grow custom installers.

--

I think this looks pretty nice, but I'm leaving it open for more comments and testing.

One thing we could also be added would be automatically detecting `layout=uki` if not configured and the positional argument is an UKI.

**Comment by behrmann** [comment]
> > I guess it would be clearer to just drop everything but `KERNEL_INSTALL_STAGING_AREA` from where to read UKIs - at least for now.
> 
> No, please don't do this. Packaged UKIs are already appearing (see the linked copr repo in the abovementioned proposal), and I don't want people to grow custom installers.

I don't have particularly strong feelings either way. The thought there was mostly to nudge people into using the staging area and have some other `.install` file put the UKI into the staging area when shipped by the distro, which I think would make it unnecessary to pass it on the command line (I haven't yet looked at the Fedora proposal). Since the initrd that `.install` scripts receive as cli argument will nowadays most likely be generated locally, I guess it should not really be a problem to change that argument to initrd or UKI.

**Comment by poettering** [comment]
Hmm, maybe my concerns regarding the kernel-install parameter either having to be a UKI or not would be alleviated if we strictly disable the plugin unless the kernel suffix is not ".efi". And then dictate that UKIs have to be suffixed with ".efi", and traditional kernels must not be.

Maybe this is something to document in the UKI spec btw, that the suffix should really be ".efi". I mean, effectively it has to be anyway, but we probably should say this literally.

File: src/kernel-install/91-ukiinstall.install
Comment (poettering) [inline src/kernel-install/91-ukiinstall.install]: hmm?

File: src/kernel-install/91-ukiinstall.install
Comment (poettering) [inline src/kernel-install/91-ukiinstall.install]: whta's the rationale for covering multiple here?

shouldn't it be simpler? i.e. look for exactly one fixed file in the staging area ("uki.efi" or so), and use that. if it doesn't exist, use the kernel passed on the kernel-install command line, but insist it is suffixed with ".efi". the net result would be: for each "kernel-install add" invocation we'd install exactly one kernel image?

File: src/kernel-install/91-ukiinstall.install
Comment (poettering) [reply src/kernel-install/91-ukiinstall.install]: copypasta?

Comment (behrmann) [reply src/kernel-install/91-ukiinstall.install]: fixed

File: src/kernel-install/91-ukiinstall.install
Comment (behrmann) [reply src/kernel-install/91-ukiinstall.install]: changed to try `uki.efi` in the staging area first and then what was passed in as kernel image.

**Comment by behrmann** [comment]
At least in BLS it already says for type 2
> the files must have the extension .efi

but I reworked the manpage to mention the new `.install` file. PTAL.

**Comment by poettering** [comment]
lgtm, but i'll leave the final review to @keszybz as he's much more of a shell pro than I am.

**Comment by bluca** [comment]
`91-ukiinstall.install` is a bit of a mouthful, what about just `91-uki.install`?

**Comment by behrmann** [comment]
Sure, I find the double-i a bit difficult. I think it was the first that @keszybz thought of on the spot. We could also lower the number to `90-uki.install`, to put it on the same footing as `90-loderentry.install`, since both are mutually exclusive - the former only acts on layout `uki` the latter only on `bls`, but I'll leave the final say to @keszybz.

**Comment by poettering** [comment]
so i'd guess sooner or later we'll have "85-uki-build.install" or so. Hence maybe "90-uki-copy.install" for this script?



("copy" instead of "install" simply because "…install.install" is just weird...

**Comment by 05storm26** [comment]
> so i'd guess sooner or later we'll have "85-uki-build.install"

Isn't that just gonna be done by the current `mkinitcpio.install` files shipped by distros/initramfs generators based on what the value of `KERNEL_INSTALL_LAYOUT`?

**Comment by behrmann** [comment]
> so i'd guess sooner or later we'll have "85-uki-build.install" or so. Hence maybe "90-uki-copy.install" for this script?
> 
> ("copy" instead of "install" simply because "…install.install" is just weird...

Changed the name as suggested due to popular demand :)

**Comment by behrmann** [comment]
> > so i'd guess sooner or later we'll have "85-uki-build.install"
> 
> Isn't that just gonna be done by the current `mkinitcpio.install` files shipped by distros/initramfs generators based on what's the value of `KERNEL_INSTALL_LAYOUT`?

There's a shared script in the making in #25180, so that this functionality does not need to be reimplemented in every generator.

**Review by bluca — APPROVED** [review]

File: src/kernel-install/meson.build
Comment (keszybz) [inline src/kernel-install/meson.build]: `uki_copy_install`

File: src/kernel-install/90-uki-copy.install
Comment (keszybz) [inline src/kernel-install/90-uki-copy.install:71]: Do we want to force `uki.efi` or should we look up any name ending in `.efi`? I guess it's OK to require a fixed name for now, we can always relax it later if needed.

**Review by keszybz — COMMENTED** [review]
LGTM. I think we should merge this. We have some time until next release, so it's better to merge things and try to make use of those features in other places (e.g. in package installation scriptlets) to see if anything is missing or needs adjusting.

File: src/kernel-install/meson.build
Comment (behrmann) [reply src/kernel-install/meson.build]: changed.

File: src/kernel-install/90-uki-copy.install
Comment (behrmann) [reply src/kernel-install/90-uki-copy.install:71]: @poettering suggested that earlier. Globbing on `*.efi` does have the downside that it leaves the case with multiple UKIs in the staging area open so that some logic is needed to decide which one to take and/or to warn in that case. I think it's easier to be strict now, though there might be a case for multiple UKIs (and initrds and loader entries for `bls`) created on a single `kernel-install` invocation. I think mkinitcpio wants to create both a normal and a fallback initrd and in Debian [a discussion is starting](https://lists.debian.org/debian-kernel/2022/11/msg00012.html) where @bluca and I are arguing in favour of `kernel-install` and maybe some added needs will be articulated there.

**Pull request merged**
